File: 073_set_matrix_zeroes/Solution2/Solution.go
package main

import "fmt"

func setZeroes(matrix [][]int)  {
    colFlag := false
    m, n := len(matrix), len(matrix[0])

    for i := 0; i < m; i++ {
        if matrix[i][0] == 0 { colFlag = true }
        for j := 1; j < n; j++ {
            if matrix[i][j] == 0 {
                matrix[0][j] = 0
                matrix[i][0] = 0
            }
        }
    }

    for i := m - 1; i >= 0; i-- {
        for j := n - 1; j > 0; j-- {
            if matrix[i][0] == 0 || matrix[0][j] == 0 {
                matrix[i][j] = 0
            }
        }
        if colFlag { matrix[i][0] = 0 }
    }
}

func main() {
    matrix := [][]int {
        {0, 1, 2, 0},
        {3, 4, 5, 2},
        {1, 3, 1, 5},
    }
    setZeroes(matrix)

    for _, row := range matrix {
        fmt.Println(row)
    }
}

File: 073_set_matrix_zeroes/Solution2/Solution.py
class Solution:
    def setZeroes(self, matrix: list) -> None:
        col_flag = False
        m, n = len(matrix), len(matrix[0])

        for i in range(m):
            if matrix[i][0] == 0: col_flag = True
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = matrix[0][j] = 0

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, 0, -1):
                if (matrix[i][0] == 0 or matrix[0][j] == 0):
                    matrix[i][j] = 0
            if col_flag: matrix[i][0] = 0



if __name__ == '__main__':
    matrix = [
        [0,1,2,0],
        [3,4,5,2],
        [1,3,1,5]
    ]
    Solution().setZeroes(matrix)

    for row in matrix:
        print(row)

